src/stan_math_backend/Expression_gen.ml

```diff
@@ -229,6 +229,10 @@ and pp_scalar_binary ppf op fn es =
   else pp_binary_f ppf fn es
 
 and gen_operator_app op ppf es =
+  let remove_basic_promotion (e : 'a Expr.Fixed.t) =
+    match e.pattern with Promotion (e, _, _) when is_scalar e -> e | _ -> e
+  in
+  let es = List.map ~f:remove_basic_promotion es in
```

Comment:
Are you sure removing promotions is always ok? Is it possible to have a division operator applied to ints-promoted-to-reals?

Reply:
The scalar signatures defined for divide are:

```ocaml
  add_unqualified ("divide", ReturnType UComplex, [UComplex; UComplex], AoS) ;
  add_unqualified ("divide", ReturnType UInt, [UInt; UInt], SoA) ;
  add_unqualified ("divide", ReturnType UReal, [UReal; UReal], SoA) ;
```
Because `int, int => int` is already a signature, that case will not generate any promotions. The only promotions that could exist are if one of the two arguments is real or complex, in which case C++ will do the right thing

To put it another way: the chunk of code you're highlighting here was actually added in order to make the output C++ consistent with what currently happens. Without it, the compiler codegens many more calls to `promote_scalar` than it did before

Reply:
Right, C++ promotions agree with Stan frontend so it should always work.
But optimzation can break this.
```stan
model {
  real bad_value = 1;
  target += bad_value / 2;
}
```
optimizes to
```
log_prob {
  {
    data real bad_value;
    target += (promote(1, real) / promote(2, real));
  }
}
```
which needs the promotions.